[package] mangle imported module names

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50049 [package] mangle imported module names**

Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

suo committed:
[package] mangle imported module names

Rationale and implementation immortalized in a big comment in
_mangling.py

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
_mangling.py

[ghstack-poisoned]

zdevito commented on torch/package/_mangling.py:
```diff
@@ -0,0 +1,140 @@
+        # Check for whether this demangled name collides with a previously saved module.
+        existing_name = self.demangled_names.get(demangled)
+        if existing_name is None:
+            self.demangled_names[demangled] = name
```
Putting state into the demangler is surprising. I can imagine situations where someone doesn't realize this object is stateful and then makes a copy of it, breaking the checks. It would be better for this logic to be handled in the export process which is already making sure a package is only exported the first time it is seen.

zdevito commented on torch/package/_custom_import_pickler.py:
```diff
@@ -54,6 +55,11 @@ def save_global(self, obj, name=None):
         lastname = name.rpartition('.')[2]
         if parent is module:
             name = lastname
+
```
I think the right place to demangle is in `whichmodule`-> it should just remove the demangled name. Then `import_module` will resolve it to the first namespace which has the unmangled thing provided. The error reporting in the pickler should be improved so that when something doesn't match it reports that one was loaded from a package when it occurs.

zdevito commented on torch/package/_custom_import_pickler.py:
```diff
@@ -96,11 +102,11 @@ def import_module_from_importers(module_name, importers):
         raise ModuleNotFoundError(module_name)
 
-def create_custom_import_pickler(data_buf, importers):
+def create_custom_import_pickler(data_buf, importers, demangler):
```
I am wary of passing the demangler here because the fast path doesn't use it.

zdevito commented on torch/package/exporter.py:
```diff
@@ -73,6 +74,8 @@ def __init__(self, filename: str, verbose: bool = True):
         self.importers = [importlib.import_module]
         self.patterns : List[Tuple[Any, Callable[[str], None]]] = []  # 'any' is 're.Pattern' but breaks old mypy
         self.debug_deps : List[Tuple[str, str]] = []
+        self.written_files : Set[str] = set()
```
This is unused

suo replied:
oops

zdevito commented on torch/package/importer.py:
```diff
@@ -239,6 +253,12 @@ def _gcd_import(self, name, package=None, level=0):
         the loader did not.
 
         """
+        if name == self._mangler.parent_name():
```
I don't get this part -- something is wrong if the import system is trying to import a package with the mangled prefix still attached. This indicates that something didn't demangle a name before using it.

suo replied:
hm yeah—my initial thought was that `_gcd_import` was the right place to put the mangling logic (since it is the greatest common denominator of all the import pathways) but I think it's actually more confusing compared to demangling at the highest level.

zdevito reviewed (commented):
I'd suggest taking the approach of trying to unmangle and mangle the names at the earliest possible point in the process of handling them. Several cases (like trying to resolve a package for __torch_package_0) are probably happening because the mangled prefix hasn't be removed before loading something. The _only_ place mangled names should exist is in the __name__ field of a module and the __module__ field of classes. When those get loaded for use in our package infrastructure they should the be removed right away. Error reporting could further look into what the mangled name was to described what went wrong.

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
_mangling.py

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
_mangling.py

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
_mangling.py

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
_mangling.py

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
_mangling.py

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

zdevito commented on torch/package/mangling.md:
```diff
@@ -0,0 +1,101 @@
+    - `__module__`
+    - `__file__`
+2. No mangled names should be serialized by `PackageExporter`.
+3. Every entry point to `PackageImporter` and `PackageExporter` must
```
I feel like this rule should be that you must pass demanded names to the package importer and exporter. The surface area of the importer/exporter API is to large to try to maintain this invariant.

zdevito commented on torch/package/exporter.py:
```diff
@@ -198,6 +228,7 @@ def _get_source_of_module(self, module: types.ModuleType) -> str:
         return ''.join(result)
 
     def require_module_if_not_provided(self, module_name: str, dependencies=True):
+        module_name = self._demangle(module_name)
```
I feel like this demangling is coming too late. The only mangled properties are the `__name__` fields of modules. I feel like the only safe place to put a demangle call is directly in the same function that looks up `__name__` from the module. Otherwise, you get this big separation between when a mangled name is extract and when it is used. The _only_ place we should want mangled names is in the `__name__` field so that Python infra doesn't get confused about module lookup.

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

zdevito commented on torch/package/_custom_import_pickler.py:
```diff
@@ -36,9 +41,15 @@ def save_global(self, obj, name=None):
+                msg = f"Can't pickle {obj}: it's not the same object as {obj2_parent_module_name}.{name}."
+
+                if is_mangled(getattr(obj, "__module__", "")) or is_mangled(obj2_parent_module_name):
+                    msg += ("\nThis is likely due to a dependency collision in torch.package; "
```
nit: maybe an explanation of how to resolve the problem 'namespace defined by package.importers resolves {untangled_name} to ... but the object is coming from ..., a module loaded from a torch.package. If you want to save this object you need to add the importer for the package to the list of importers ..."

zdevito commented on torch/package/_mangling.py:
```diff
@@ -0,0 +1,52 @@
+    def mangle(self, name) -> str:
+        return self._mangle_parent + "." + name
+
+    def demangle(self, mangled) -> str:
```
I am not sure this version is really needed and which one of the demangles to use might confuse people working with this code. I think it is safe to always use the global demangle since the only case where we use this one is `_get_package` where we know already that the name is mangled locally.

zdevito commented on torch/package/exporter.py:
```diff
@@ -162,8 +164,15 @@ def save_source_string(self, module_name: str, src: str, is_package: bool = Fals
+        except ModuleNotFoundError as e:
+            if not is_mangled(module_name):
+                raise
+            msg = (f"Cannot export module '{module_name}'. Modules imported "
```
Probably should still say "module not found". It's name has been mangled because it was loaded from a torch package. If you want to export it ...

zdevito reviewed (approved):
Looks good! Only minor comments.

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]

suo committed:
Update on "[package] mangle imported module names"


Rationale and implementation immortalized in a big comment in
`torch/package/mangling.md`.

This change also allows imported modules to be TorchScripted

Differential Revision: [D25758625](https://our.internmc.facebook.com/intern/diff/D25758625)

[ghstack-poisoned]
